qiskit/transpiler/passes/adjacent_barriers.py

```diff
@@ -0,0 +1,120 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2019, IBM.
+#
+# This source code is licensed under the Apache License, Version 2.0 found in
+# the LICENSE.txt file in the root directory of this source tree.
+"""
+A pass that merges any adjacent barriers into one
+"""
+
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.dagcircuit import DAGCircuit
+from qiskit.extensions.standard.barrier import Barrier
+
+
+class MergeAdjacentBarriers(TransformationPass):
+    """Returns a circuit with any adjacent barriers merged together"""
+
+    def run(self, dag):
+
+        # sorted to so that they are in the order they were added to the DAG
+        # so ancestors/descendants makes sense
+        barriers = sorted(dag.named_nodes('barrier'))
+
+        # get dict of barrier merges
+        node_to_barrier = MergeAdjacentBarriers._collect_potential_merges(dag, barriers)
+
+        if not node_to_barrier:
+            return dag
+
+        # add the merged barriers to a new DAG
+        new_dag = DAGCircuit()
+
+        for qreg in dag.qregs.values():
+            new_dag.add_qreg(qreg)
+        for creg in dag.cregs.values():
+            new_dag.add_creg(creg)
+
+        # go over current nodes, and add them to the new dag
+        for node in dag.nodes_in_topological_order():
+
+            if node.type == 'op':
+                if node.name == 'barrier':
+                    if node in node_to_barrier:
+                        new_dag.apply_operation_back(node_to_barrier[node])
+                else:
+                    # copy the condition over too
+                    if node.condition:
+                        new_dag.apply_operation_back(node.op, condition=node.condition)
+                    else:
+                        new_dag.apply_operation_back(node.op)
+        return new_dag
+
+    @staticmethod
+    def _collect_potential_merges(dag, barriers):
+        """
+        Returns a dict of DAGNode : Barrier objects, where the barrier needs to be
+        inserted where the corresponding DAGNode appears in the main DAG
+        """
+        # if only got 1 or 0 barriers then can't merge
+        if len(barriers) < 2:
+            return None
+
+        # mapping from the node that will be the main barrier to the
+        # barrier object that gets built up
+        node_to_barrier = {}
+
+        # Start from the first barrier
+        current_barrier = barriers[0]
+        end_of_barrier = current_barrier
+        current_barrier_nodes = [current_barrier]
+
+        current_qubits = set(current_barrier.qargs)
+        current_ancestors = dag.ancestors(current_barrier)
+        current_descendants = dag.descendants(current_barrier)
+
+        barrier_to_add = Barrier(qubits=current_qubits)
+
+        for next_barrier in barriers[1:]:
+
+            next_ancestors = {nd for nd in dag.ancestors(next_barrier)
```

Comment:
Can you add a comment clarifying why `if nd not in current_barrier_nodes` is important? Alternately, you could update e.g. the `current_ancestors` update to be something like `(current_ancestors | next_ancestors) - {next_barrier}`.